flybywiresim/a32nx
# feat: Flypadosv3 - Presets for Lighting and Aircraft

## Targets flypadosv3

Put these into the flypadosv3 PR when merged: #6230, 

## Summary of Changes
This PR implements presets for interior lighting and for aircraft state.

Interior Lighting Presets: User can load and save the state of interior lighting setting via the EFB.

Aircraft States; The user can choose 5 presets from the EFB which will be applied to the aircraft. The presets will be applied by executing the correct procedure as if done by a Co-Pilot. The presets are therefore NOT applied instantaneous.

### Architecture
#### Front End:
EFB presets page with tabs for lighting and aircraft presets (see screenshots)
Uses  LVARs to request the presets being applied by the wasm backend. 

#### Backend:
C++ WASM backend to overcome MSFS JS limitation to execute certain tasks (calculator codes).
The backend checks the request LVARs and applies the requested preset settings accordingly. The request LVAR is reset to 0 after applying the preset. 

Lighting:
- For lighting this is instantaneous.
- Lighting settings are stored in an 'ini' in the /work/ folder next to the ThrottleCalibration.ini, etc.
- Default values of 50% for all lights are applied when a new preset ID is loaded

Aircraft:
- For aircrafts presets this is done in real time by executing predefined procedure steps in real time and waiting for certain conditions if necessary (e.g. APU Start, Engine Start, etc.). 
- The procedure presets are hard coded in a simple data structure in the C++ (header file) using MSFS Calculator Code (RNP notation). This could eventually also move to a configuration file if users are requesting it.
- The procedure steps check if a step need executing at all or if its state is already set
- Also special steps to wait for specific state are possible (e.g. wait for APU Start, Engine Start, ADIRS align, etc.)

#### LVARs:
- LVAR: A32NX_LOAD_LIGHTING_PRESET
  - value; ID for the requested preset, 0 when no request is active
- LVAR: A32NX_SAVE_LIGHTING_PRESET
  - value; saves current lighting settings to the preset slot provided presets slot ID , 0 when no request is active or saving is done
- LVAR: A32NX_LOAD_AIRCRAFT_PRESET
  - value; ID for the requested preset, 0 when no request is active or when loading shall be stopped
- LVAR: A32NX_LOAD_AIRCRAFT_PRESET_PROGRESS
  - value: percentage of steps done while loading a preset. 

## Screenshots
Layout might still change.

Lighting Presets Page:
![image](https://user-images.githubusercontent.com/16833201/158632832-ec820412-0bf2-4515-876f-34556b8d9633.png)

Aircraft Presets Page:
![image](https://user-images.githubusercontent.com/16833201/158633011-f6219fe4-8ed0-485a-99a1-0141511d43d7.png)
![image](https://user-images.githubusercontent.com/16833201/158633126-32b711a8-2d77-433a-a077-af9b36079c4d.png)

Video:
https://user-images.githubusercontent.com/16833201/158637942-5592cd6e-20fe-49a5-be83-7599b3213cfa.mp4


Discord username (if different from GitHub): Cdr_Maverick#6475

## Testing instructions
### General
- Loading of the aircraft - this PR adds another wasm file. I would be interested in feedback how much difference this makes for you startup time after loading the PR for the first time.
- Feedback on the layout of the preset pages in general 
	- is it easy to understand?
	- what could be improved?

### Lighting
- Presets cannot be loaded or saved when aircraft is without power
- After turning on Ext Pwr, APU or Engines presets should be loadable/saveable
- Test changing of preset names (e.g. test none ASCII characters, etc.)
- Test saving & laoding of presets - check that all settings are save and correctly loaded again
- Check if presets are correctly saved to ini file: 
  - MS STORE: C:\Users\USERNAME\AppData\Local\Packages\Microsoft.FlightSimulator_8wekyb3d8bbwe\LocalState\packages\flybywire-aircraft-a320-neo\work\InteriorLightingPresets.ini
- Change values in the ini and check that they are loaded correctly
- Try to break the ini file

### Aircraft Presets
- Test loading of presets from C&D
- Test loading a preset twice - no action should be executed
- Test jumping between presets - everything should be setup correctly
- Test changing some settings manually then load a preset - everything should be setup correctly
- Test canceling loading - loading should immediately stop. Procedure will be incomplete.
  - Load the same preset or another to make sure the procedure is complete - check if setup is correct
- Test starting at gate or runway
- Try setting a preset while in the air - should not be possible - only allowed on ground
- Try breaking it in any way 😉 

### Help page 
- Check if understandable
- Check for typos


## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created, built, and uploaded.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the left side, click on the bottom **PR** tab
1. Click on the **A32NX** download link at the bottom of the page
